### PR TITLE
Build non-static binaries with PIE buildmode

### LINKED ISSUE #90311: Build position independent executables
## What would you like to be added:

Currently `kubelet` is dynamically linked but is not built as a position independent executable. It would be great if it could be built with Go’s `-buildmode=pie` flag to enable ASLR.

## Why is this needed:

I’m not a security expert, and I’m not familiar with those concepts, but it seems that ASLR is a pretty common (and recommended) security mechanism.

If I understand correctly, statically linked binaries (e.g. all the Kubernetes binaries except `kubelet`) do not require/support PIE so this would only apply to `kubelet`.

## Related issues/PR:

There was an attempt at using Go’s `-buildmode=pie` in #57225 but it seems that it was abandoned due to an incompatibility issue with Windows.

### PR DESCRIPTION
**Pull request by saschagrunert**
#### What type of PR is this?

/kind cleanup

#### What this PR does / why we need it:

We now add the `-buildmode pie` flag when building non-static binaries, which enables the ASLR security mechanism.

#### Which issue(s) this PR fixes:

Fixes #90311

#### Special notes for your reviewer:

Supersedes #94448

For example, we now build the `kubelet` as `LSB pie executable`:
```
> make WHAT=cmd/kubelet
> file _output/bin/kubelet
_output/bin/kubelet: ELF 64-bit LSB pie executable, x86-64, version 1 (SYSV), dynamically linked, interpreter /nix/store/vaf92wc23m067q9aig6zjkcnjvrg768n-glibc-2.32-46/lib/ld-linux-x86-64.so.2, for GNU/Linux 2.6.32, Go BuildID=2XJTWqoOSzrJFsezWtNV/Ljlb-W11uIqFXSJ5PWOt/4aLSZPegz6V8zUDouB74/B8q7ucGO132SbpmNnZMA, stripped
```

We do not apply the change to statically linked binaries, because if we add the `pie` buildmode to them we would result in gaining a dynamically linked binary without choosing an external linker. Ref: https://github.com/golang/go/issues/40719

#### Does this PR introduce a user-facing change?

```release-note
Changed buildmode of non static Kubernetes binaries to produce position independent executables (PIE).
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

```docs
None
```

/area release-eng
/sig release
/priority important-soon
/cc @kubernetes/release-managers @abhay-krishna

### CODE REVIEW

**Review by xmudrii — APPROVED** [review]
/lgtm

**Comment by BenTheElder** [comment]
Have we investigated fully e.g.:
- binary size impact
- Platform support? I recall this being restricted to certain platforms in go.
/test pull-kubernetes-cross

I'm still not clear on if we think this will realistically defend against anything or check a compliance checkbox. If it's only the latter then I think the GOFLAGS support and custom builds should cover that angle.

**Comment by saschagrunert** [comment]
#### Binary size comparison

Only a few binaries are build with PIE if we run `make`:

```
> file _output/bin/* | rg pie
_output/bin/gendocs:                  ELF 64-bit LSB pie executable, x86-64, version 1 (SYSV), dynamically linked, interpreter /nix/store/vaf92wc23m067q9aig6zjkcnjvrg768n-glibc-2.32-46/lib/ld-linux-x86-64.so.2, for GNU/Linux 2.6.32, Go BuildID=_d2KgGty1Y6_MSazK5Jk/3yF_B3YODNHxJQQcLO6A/QjAnNuKVWj4DboY948u2/wsZTn89yQHMAgYLlfRzL, stripped
_output/bin/genkubedocs:              ELF 64-bit LSB pie executable, x86-64, version 1 (SYSV), dynamically linked, interpreter /nix/store/vaf92wc23m067q9aig6zjkcnjvrg768n-glibc-2.32-46/lib/ld-linux-x86-64.so.2, for GNU/Linux 2.6.32, Go BuildID=pUqkV9w3NiWUQeP5VG_T/sIolGFDiizX8-m5bRW4Q/nlEHNoxpLXMCRlOpc4tR/CczSf5vapfKrHp7SzQmj, stripped
_output/bin/genman:                   ELF 64-bit LSB pie executable, x86-64, version 1 (SYSV), dynamically linked, interpreter /nix/store/vaf92wc23m067q9aig6zjkcnjvrg768n-glibc-2.32-46/lib/ld-linux-x86-64.so.2, for GNU/Linux 2.6.32, Go BuildID=xOy3GkYhsocVP1XlpjqM/V5QA7seFrY8jYQu_-2z5/cUhOL6ZFahsKsht6PWPj/hZHSFVhliVgSqi7FC3sl, stripped
_output/bin/genyaml:                  ELF 64-bit LSB pie executable, x86-64, version 1 (SYSV), dynamically linked, interpreter /nix/store/vaf92wc23m067q9aig6zjkcnjvrg768n-glibc-2.32-46/lib/ld-linux-x86-64.so.2, for GNU/Linux 2.6.32, Go BuildID=7_02dm6WLh8fPyXGlX2E/P1WpfpSt7xBaUnhRM7vk/EGkqZ4LE0HlINKbKaRID/4o5VmlZrDHg70c1Uiccs, stripped
_output/bin/kubectl-convert:          ELF 64-bit LSB pie executable, x86-64, version 1 (SYSV), dynamically linked, interpreter /nix/store/vaf92wc23m067q9aig6zjkcnjvrg768n-glibc-2.32-46/lib/ld-linux-x86-64.so.2, for GNU/Linux 2.6.32, Go BuildID=2rpVm1qyEjM93PNvoCuC/6eAjhy5q5ycwAmUnUrUU/aBpLHa02qan5X5T3IEul/BpxL3sumz7xd0aVczoEj, stripped
_output/bin/kubelet:                  ELF 64-bit LSB pie executable, x86-64, version 1 (SYSV), dynamically linked, interpreter /nix/store/vaf92wc23m067q9aig6zjkcnjvrg768n-glibc-2.32-46/lib/ld-linux-x86-64.so.2, for GNU/Linux 2.6.32, Go BuildID=Hv1ij3pn7pWX7UAeGV92/Ljlb-W11uIqFXSJ5PWOt/4aLSZPegz6V8zUDouB74/w4DzDHeUwrplTcYeo3P-, stripped
_output/bin/kubemark:                 ELF 64-bit LSB pie executable, x86-64, version 1 (SYSV), dynamically linked, interpreter /nix/store/vaf92wc23m067q9aig6zjkcnjvrg768n-glibc-2.32-46/lib/ld-linux-x86-64.so.2, for GNU/Linux 2.6.32, Go BuildID=f8w_af24P1DYtMUkIRVm/F9gwmSOF-a76VYIsHr84/KkyzMvyFIH9RKxoLDJuV/XZIL5rzhDzSWo0LnnofR, stripped
```

If we compare their binary sizes, then we indeed encounter a difference:

| Binary | without PIE (before) | with PIE (after) |
|-|-|-|
| gendocs                  | 55Mi  |  64Mi |
| genkubedocs              | 189Mi  | 226Mi |
| genman                   | 197Mi  | 235Mi|
| genyaml                  |  55Mi  |  64Mi|
| kubectl-convert          |  53Mi  |  62Mi|
| kubelet                  | 113Mi  | 143Mi|
| kubemark                 | 110Mi  | 139Mi|


#### Platform support

We have PIE mode since:
- go 1.6 for: android/386, android/amd64, android/arm, android/arm64, linux/386, linux/amd64, linux/arm, linux/arm64, and linux/ppc64le
- go 1.10 for: darwin/amd64 (also forces the use of external linking on all systems)
- go 1.13 for: aix/ppc64
- go 1.15 for: windows (on per default),  linux/amd64 and linux/arm64 defaults to internal linking mode for -buildmode=pie
- go 1.16 for: darwin/arm64

**Comment by xmudrii** [comment]
> I'm still not clear on if we think this will realistically defend against anything or check a compliance checkbox. If it's only the latter then I think the GOFLAGS support and custom builds should cover that angle.

The issue (#90311) has some context about this. This comment from @tabbysable and the following comments provide some insights why this would be useful: https://github.com/kubernetes/kubernetes/issues/90311#issuecomment-824945849

From @saschagrunert's table above, the binary size increase is about 10-20% depending on the binary. This might sound like a lot, but most of the binaries are somewhat small. I don't have experience with this concrete topic, so I can't weigh is it worth or not, but in my opinion, I don't really see this as a problem.

**Comment by BenTheElder** [comment]
I wouldn't call > 100mb small and I'm not sure many other people would either? Most of our binaries are sadly over 100mb easily. They're not all in the table above.

We can ignore the gen* as we don't distribute those.

Making kubelet ~27% larger is not great :(

**Comment by BenTheElder** [comment]
This broke the cross build
/hold

**Comment by saschagrunert** [comment]
Looks like we're missing `gcc-multilib` package for linux/386 to link the client binaries.

File: hack/lib/golang.sh
Comment (saschagrunert) [inline hack/lib/golang.sh:418]: We require a new kube-cross version for this, see: https://github.com/kubernetes/release/pull/2086

**Comment by saschagrunert** [comment]
The new kube-cross version makes `pull-kubernetes-cross` green. PTAL again @BenTheElder

**Comment by BenTheElder** [comment]
> Depending on how the compiler behaves and the output of pull-kubernetes-cross, we may have to exclude architectures like s390x.

It seems that wasn't the case. https://prow.k8s.io/view/gs/kubernetes-jenkins/pr-logs/pull/102323/pull-kubernetes-cross/1397898981865951232

As an aside: I wonder how long the project should keep building for relatively obsolete platforms like i386 ...

Still disappointed to see kubelet get bigger again but oh well. https://groups.google.com/g/golang-nuts/c/cXhRsmNsMwo/m/VUZlLqo9AwAJ seems pretty clear that there's at least some benefit to programs using CGO (like kubelet to an extent, indirectly).

I also asked the GKE security team for more background on this since there's been pretty limited discussion on github and I've already poked sig-security, they had a similar response to the effect of roughly "it's probably relatively minor / theoretical but perhaps useful with CGO".

in either case, the build script changes look acceptable purely from a code POV without respect to "should we PIE or not" and PIE seems at least _potentially_ useful for these non-static binaries.

/approve

and since cross is green again:
/hold-cancel

**Comment by saschagrunert** [comment]
Referring to: https://github.com/kubernetes/kubernetes/pull/102323#issuecomment-849934751
/hold cancel

**Comment by saschagrunert** [comment]
@kubernetes/release-engineering @kubernetes/sig-security-pr-reviews please take a look for lgtm

**Pull request merged**
